packages/lesswrong/components/hooks/useCookiesWithConsent.ts

```diff
@@ -0,0 +1,87 @@
+  const cookiePreferences = explicitConsentGiven ? preferencesCookieValue : fallbackPreferences
+
+  // If we can't determine whether explicit consent is required synchronously (from localStorage), check via the geolocation API
+  useEffect(() => {
+    if (explicitConsentRequired !== "unknown") return;
+
+    void (async () => {
+      const explicitConsentRequired = await getExplicitConsentRequiredAsync();
+      setExplicitConsentRequired(explicitConsentRequired);
+    })();
+  }, [explicitConsentRequired]);
+
+  // If the user had not given explicit consent, but the value of COOKIE_PREFERENCES_COOKIE is different to what we are
+  // using in the code (fallbackPreferences), update the cookie. This is so that Google Tag Manager handles it correctly.
+  useEffect(() => {
+    if (explicitConsentRequired === "unknown" || explicitConsentGiven) return;
+
+    if (JSON.stringify(cookiePreferences) !== JSON.stringify(preferencesCookieValue)) {
+      setCookie(COOKIE_PREFERENCES_COOKIE, cookiePreferences, { path: "/" });
+      void cookiePreferencesChangedCallbacks.runCallbacks({iterator: cookiePreferences, properties: []});
+    }
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [explicitConsentRequired, JSON.stringify(cookiePreferences), JSON.stringify(preferencesCookieValue), setCookie]);
+  
+  const updateCookiePreferences = useCallback(
+    (newPreferences: CookieType[]) => {
+      captureEvent("cookiePreferencesUpdated", {
+        cookiePreferences: newPreferences,
+      })
+      setCookie(COOKIE_CONSENT_TIMESTAMP_COOKIE, new Date(), { path: "/" });
+      setCookie(COOKIE_PREFERENCES_COOKIE, newPreferences, { path: "/" });
+      void cookiePreferencesChangedCallbacks.runCallbacks({iterator: newPreferences, properties: []});
+    },
+    [captureEvent, setCookie]
+  );
+  return { cookiePreferences, updateCookiePreferences, explicitConsentGiven, explicitConsentRequired };
+}
+
+export function useCookiesWithConsent(dependencies?: string[]): [
+  {
+    [name: string]: any;
+  },
+  (name: string, value: any, options?: CookieSetOptions) => void,
+  (name: string, options?: CookieSetOptions) => void
+] {
+  const { cookiePreferences } = useCookiePreferences();
+
+  const [cookies, setCookieBase, removeCookieBase] = useCookies(dependencies);
+
+  const setCookie = useCallback(
+    (name: string, value: string, options?: CookieSetOptions) => {
+      if (!isCookieAllowed(name, cookiePreferences)) {
+        // eslint-disable-next-line no-console
+        console.warn(`Consent has not been granted for cookie "${name}" to be set`);
+        return;
+      }
+
+      setCookieBase(name, value, options);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [JSON.stringify(cookiePreferences), setCookieBase]
+  );
+
+  return [cookies, setCookie, removeCookieBase];
+}
```

Comment:
This function only filters on consent when **setting** a cookie, but shouldn't we be removing or ignoring existing cookies if the user changes their preferences? In particular, if a user has already been using the site and then clicks "Reject" on the cookie banner, it seems like we should no longer be using non-necessary cookies for them.

Reply:
I've added this bit of copy to the dialog and the policy to cover this:
```
If you choose to reject cookies, you are responsible for removing any that have already been set (such as if
you previously accepted, or you visited before our cookie policy was implemented). See the instructions for
doing so <a href="https://support.google.com/chrome/answer/95647">here</a>.
```
Unfortunately it's now quite verbose

Reply:
Oh wow that sucks, I didn't know that rejecting cookies only affects ones that might be added in the future.

So I don't think that this will affect many users (and I personally don't really care about sites using cookies), but I still think that this is misleading, and it's not that hard to make it work as a user would probably expect. Ex. we have the consent data in here, could we filter the returned list of `cookies` based on that? Or clear the proper cookies in `updateCookiePreferences()`?

Then you could remove the added paragraph and make the dialog look neater :)

Reply:
I've added a callback which removes any disallowed cookies when they change their preferences, this means it will work for users who have never accepted cookies. Unfortunately datadog constantly re-sets it's cookie and can't be disabled once it's initialised, so for people who have accepted and then reject this won't 100% work. I've changed the text to this to cover this:
```
If you have previously accepted cookies and are now rejecting them, you are responsible for removing any that
have already been set. You can do so by refreshing the page and then following the instructions{" "}
<a href="https://support.google.com/chrome/answer/95647">here</a>.
```